### PR TITLE
Fix b/33924233 Use noIndex instead of notFound when skipping docs

### PR DESCRIPTION
**Pull request by BrettMichaelJohnson**
This change has a TransmissionDecision of do-not-index set
the "X-Robots-Tag=noindex" header, rather than returning a 404
response code.

### CODE REVIEW

**Comment by wiarlawd** [comment]
The changes to the state machine make me nervous, for two reasons. First, I don't really understand it. Second, we decided to not use setNoFollow, on a nudge, mainly because it meant doing nothing with some slight benefit. However, seeing this change, and the comments in the state engine, makes me think that when we use setNoIndex like this, we should also avoid sending metadata, ACLs, and content. That is more consistent with one of the main benefits of SkipDocumentFilter: do no bother sending things to the GSA that you don't want to index, and retains the main benefits of this change, which is simply to change the message in Index Diagnostics. I don't mind losing followed links as a result, which was a very minor point.

John L

File: src/com/google/enterprise/adaptor/DocumentHandler.java [hunk 1]
```diff
@@ -847,30 +859,28 @@ public OutputStream getOutputStream() throws IOException {
                   .createPipeline(os, originalContentType, metadata);
             }
           }
-        } else if (state == State.SEND_BODY_TRANSFORMED_TO_NOT_FOUND) {
-          log.log(Level.INFO, "changed SEND_BODY to NOT_FOUND {0}",
+        } else if (state == State.SEND_BODY_TRANSFORMED_TO_NO_INDEX) {
+          log.log(Level.INFO, "changed SEND_BODY to DO_NOT_INDEX {0}",
               docId.getUniqueId());
           // Not using startSending. Instead we stop header timer, start
           // content timer, send not-found page, and setup a sink for
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/DocumentHandler.java]: Update "not-found page".

File: test/com/google/enterprise/adaptor/DocumentHandlerTest.java [hunk 2]
```diff
@@ -516,14 +516,14 @@ public void getDocContent(Request request, Response response)
         .build();
     mockAdaptor.documentBytes = new byte[] {1, 2, 3};
     handler.handle(ex);
-    assertEquals(404, ex.getResponseCode());
-    assertEquals(null,
-         ex.getResponseHeaders().getFirst("X-Gsa-External-Metadata"));
-    assertArrayEquals("Error 404: Not Found".getBytes(), ex.getResponseBytes());
+    assertEquals(200, ex.getResponseCode());
+    assertEquals("noindex", ex.getResponseHeaders().getFirst("X-Robots-Tag"));
+    assertNull(ex.getResponseHeaders().getFirst("X-Gsa-External-Metadata"));
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/DocumentHandlerTest.java]: Revert to assertEquals as discussed.

File: test/com/google/enterprise/adaptor/DocumentHandlerTest.java [hunk 3]
```diff
@@ -550,11 +550,12 @@ public void getDocContent(Request request, Response response)
...
     handler.handle(headEx);
-    assertEquals(404, headEx.getResponseCode());
-    assertEquals(null,
-         headEx.getResponseHeaders().getFirst("X-Gsa-External-Metadata"));
-    // 404 response to HEAD request does not have content (it is HEAD)
-    assertArrayEquals("".getBytes(), headEx.getResponseBytes());
+    assertEquals(200, headEx.getResponseCode());
+    assertEquals("noindex",
+        headEx.getResponseHeaders().getFirst("X-Robots-Tag"));
+    assertNull(headEx.getResponseHeaders().getFirst("X-Gsa-External-Metadata"));
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/DocumentHandlerTest.java]: ditto

File: src/com/google/enterprise/adaptor/DocumentHandler.java [hunk 4]
```diff
@@ -847,30 +859,28 @@ public OutputStream getOutputStream() throws IOException {
...
-          countingOs = new CountingOutputStream(new CloseNotifyOutputStream(
-              ex.getResponseBody()));
-          log.finest("before writing chunked not-found response");
-          countingOs.write(
-              Translation.HTTP_NOT_FOUND.toString().getBytes(ENCODING));
-          log.finest("after writing chunked not-found response");
-          os = new SinkOutputStream(countingOs);
+          // Follow the precedent of HEAD doing a short-circuit out
+          // and potentially not reporting some later adaptor errors.
+          os = new CloseNotifyOutputStream(new SinkOutputStream());
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/DocumentHandler.java]: As discussed, preserve the call to ex.getResponseBody() as new SinkOS(new CloseNotifyOS(ex.getResponseBody())), both here and below.

**Review by wiarlawd — CHANGES_REQUESTED** [review]
As discussed.

File: src/com/google/enterprise/adaptor/DocumentHandler.java [hunk 5]
```diff
@@ -847,30 +859,26 @@ public OutputStream getOutputStream() throws IOException {
...
           // bytes provided by adaptor instance itself.
           watchdog.processingCompleted(workingThread);
           watchdog.processingStarting(workingThread, contentTimeoutMillis);
-          int rc = HttpURLConnection.HTTP_NOT_FOUND;
-          HttpExchanges.startResponse(ex, rc, "text/plain", /*hasBody=*/ true);
-          countingOs = new CountingOutputStream(new CloseNotifyOutputStream(
+          // Send no body and no metadata, other than Robots noindex.
+          ex.getResponseHeaders().add("X-Robots-Tag", "noindex");
+          int rc = HttpURLConnection.HTTP_OK;
+          HttpExchanges.startResponse(ex, rc, "text/plain", /*hasBody=*/ false);
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/DocumentHandler.java:873]: I don't really understand the point of lying and sending 0 bytes of chunked content below, but why not preserve that behavior here?

File: test/com/google/enterprise/adaptor/DocumentHandlerTest.java [hunk 6]
```diff
@@ -725,11 +728,11 @@ public void getDocContent(Request request, Response response)
...
     mockAdaptor.documentBytes = new byte[] {1, 2, 3};
     handler.handle(ex);
-    assertEquals(404, ex.getResponseCode());
+    assertEquals(204, ex.getResponseCode());
+    assertEquals("noindex", ex.getResponseHeaders().getFirst("X-Robots-Tag"));
     assertEquals(null,
-         ex.getResponseHeaders().getFirst("X-Gsa-External-Metadata"));
-    // The 204 that would not have content becomes a 404 with content
-    assertArrayEquals("Error 404: Not Found".getBytes(), ex.getResponseBytes());
+        ex.getResponseHeaders().getFirst("X-Gsa-External-Metadata"));
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/DocumentHandlerTest.java]: Maybe unfix broken indentation for git history?

**Review by wiarlawd — APPROVED** [review]
LGTM, with minor/trivial suggestions.

Please wait for PJ to review as well.

John L

File: src/com/google/enterprise/adaptor/DocumentHandler.java [hunk 7]
```diff
@@ -847,30 +859,26 @@ public OutputStream getOutputStream() throws IOException {
...
-          countingOs.write(
-              Translation.HTTP_NOT_FOUND.toString().getBytes(ENCODING));
-          log.finest("after writing chunked not-found response");
-          os = new SinkOutputStream(countingOs);
         } else if (state == State.SEND_BODY_TRANSFORMED_TO_HEAD) {
           log.log(Level.INFO, "changed SEND_BODY to HEAD {0}",
               docId.getUniqueId());
           startSending(true);  // Lie about content. Will be 0bytes chunked.
-          countingOs = new CountingOutputStream(new CloseNotifyOutputStream(
+          os = new SinkOutputStream(new CloseNotifyOutputStream(
```
Comment (PapaJoltOmega) [inline src/com/google/enterprise/adaptor/DocumentHandler.java:880]: Why this change?

File: src/com/google/enterprise/adaptor/DocumentHandler.java (hunk 5 above)
Comment (PapaJoltOmega) [reply src/com/google/enterprise/adaptor/DocumentHandler.java:873]: Chunked was intentional. I believe it was to preserve error behaviour because completed and successfully closed connection is different from one that results in 500 half way through document processing.

File: src/com/google/enterprise/adaptor/DocumentHandler.java [hunk 8]
```diff
@@ -847,30 +859,26 @@ public OutputStream getOutputStream() throws IOException {
...
               docId.getUniqueId());
           // Not using startSending. Instead we stop header timer, start
-          // content timer, send not-found page, and setup a sink for
+          // content timer, discard metadata and content, and setup a sink for
           // bytes provided by adaptor instance itself.
           watchdog.processingCompleted(workingThread);
           watchdog.processingStarting(workingThread, contentTimeoutMillis);
-          int rc = HttpURLConnection.HTTP_NOT_FOUND;
-          HttpExchanges.startResponse(ex, rc, "text/plain", /*hasBody=*/ true);
-          countingOs = new CountingOutputStream(new CloseNotifyOutputStream(
```
Comment (PapaJoltOmega) [inline src/com/google/enterprise/adaptor/DocumentHandler.java:860]: Why this change?

**Review by PapaJoltOmega — COMMENTED** [review]
Thank you for this CL. I was away last week and didn't get to review.  

Looks like this CL was submitted to master HEAD already and I'm not yet sure it is right. 

Thank you

File: src/com/google/enterprise/adaptor/DocumentHandler.java (hunk 7 above)
Comment (BrettMichaelJohnson) [reply src/com/google/enterprise/adaptor/DocumentHandler.java:880]: The CountingOuputStream provided no benefit here (and in SEND_BODY_TRANSFORMED_TO_NO_INDEX above). It is used only to journal the number of bytes of content written.  In these two cases, that number is always 0 (since we are no longer forging a 404 response page).  If countingOs is null, then the journal records 0 bytes written.

File: src/com/google/enterprise/adaptor/DocumentHandler.java (hunk 5 above)
Comment (BrettMichaelJohnson) [reply src/com/google/enterprise/adaptor/DocumentHandler.java:873]: So you concur with John, in that I should lie here with hasBody = true, even though no bytes get written?  I can do that.

Comment (PapaJoltOmega) [reply src/com/google/enterprise/adaptor/DocumentHandler.java:873]: Yes, because at this moment I believe the behaviour is different.

I think the commented behaviour on line 691 says that "use same timeouts and error handling as SEND_BODY.".  

On one hand we have: 0 sized chunked that reads entire stream before closing connection 

On other we have: 0 sized that is immediately closed

There are timing differences and there are differences when there is error in reading body.

File: src/com/google/enterprise/adaptor/DocumentHandler.java (hunk 7 above)
Comment (PapaJoltOmega) [reply src/com/google/enterprise/adaptor/DocumentHandler.java:880]: I would appreciate anything that can be done to limit this CL's modifications because changes to this file stress my mental capabilities.  Does your comment mean that removing the COuntingOutputStream wrapper is not essential to this CL and that we can consider such removal orthogonally to this CL?  I would greatly appreciate any fragment that can be extracted and seperated be done in seperate CL. Thank you

**Review by PapaJoltOmega — COMMENTED** [review]
Thank you

File: src/com/google/enterprise/adaptor/DocumentHandler.java (hunk 7 above)
Comment (wiarlawd) [reply src/com/google/enterprise/adaptor/DocumentHandler.java:880]: Fwiw, Brett removed CountingOutputStream for basically the same reason: this code is hard to reason about, with many subtle differences, and having no-op code wasn't helping us understand the subtle differences. He's going to put up a new commit here to minimize diffs, but it will look a little strange on master (assuming there are other subsequent changes), as the countingOs will be removed, added back, and then possibly removed again.

**Review by PapaJoltOmega — COMMENTED** [review]
LGTM.  Thank you.

**Comment by BrettMichaelJohnson** [comment]
This pull request was reverted, and a new one with additional changes was created [here](https://github.com/googlegsa/library/pull/22)
